statsmodels/sandbox/mice/mice.py

```diff
@@ -0,0 +1,1287 @@
+                continue
+            if (v.ndim == 1) and (v.shape[0] == len(rix)):
+                kwds[k] = v[rix]
+            if (v.ndim == 2) and (v.shape[0] == len(rix)):
+                kwds[k] = v[rix, :]
+
+        return kwds
+
+
+    def _perturb_bootstrap(self, vname):
+        """
+        Perturbs the model's parameters using a bootstrap.
+        """
+
+        endog, exog, init_kwds, fit_kwds = self.get_fitting_data(vname)
+
+        m = len(endog)
+        rix = np.random.randint(0, m, m)
+        endog = endog[rix]
+        exog = exog[rix, :]
+
+        init_kwds = self._boot_kwds(init_kwds, rix)
+        fit_kwds = self._boot_kwds(fit_kwds, rix)
+
+        klass = self.model_class[vname]
+        self.models[vname] = klass(endog, exog, **init_kwds)
+        self.results[vname] = self.models[vname].fit(**fit_kwds)
+        self.params[vname] = self.results[vname].params
+
+
+    def _perturb_gaussian(self, vname):
+        """
+        Gaussian perturbation of model parameters.
+
+        The normal approximation to the sampling distribution of the
+        parameter estimates is used to define the mean and covariance
+        structure of the perturbation distribution.
+        """
+
+        endog, exog, init_kwds, fit_kwds = self.get_fitting_data(vname)
+
+        klass = self.model_class[vname]
+        self.models[vname] = klass(endog, exog, **init_kwds)
+        self.results[vname] = self.models[vname].fit(**fit_kwds)
+
+        cov = self.results[vname].cov_params()
+        mu = self.results[vname].params
+        np.linalg.cholesky(cov) #!!!!!!!
+        self.params[vname] = np.random.multivariate_normal(mean=mu, cov=cov)
+
+
+    def perturb_params(self, vname):
+
+        if self.perturbation_method[vname] == "gaussian":
+            self._perturb_gaussian(vname)
+        elif self.perturbation_method[vname] == "boot":
+            self._perturb_bootstrap(vname)
+        else:
+            raise ValueError("unknown perturbation method")
+
+
+    def impute(self, vname):
+        # Wrap this in case we later add additional imputation
+        # methods.
+        self.impute_pmm(vname)
+
+
+    def update(self, vname):
+        """
+        Impute missing values for a single variable.
+
+        This is a two-step process in which first the parameters are
+        perturbed, then the missing values are re-imputed.
+
+        Parameters
+        ----------
+        vname : string
+            The name of the variable to be updated.
+        """
+
+        self.perturb_params(vname)
+        self.impute(vname)
+
+
+    # work-around for inconsistent predict return values
+    def _get_predicted(self, obj):
+
+        if isinstance(obj, np.ndarray):
+            return obj
+        elif hasattr(obj, 'predicted_values'):
+            return obj.predicted_values
+        else:
+            raise "cannot obtain predicted values from %s" % obj.__class__
+
+
+    def impute_pmm(self, vname):
```

Comment:
we could add here a keyword argument for a selection/aggregation function as a callback.
It can default to the current random selection, but we could use non-uniform randomization or impute the mean of the neighbors for use in propensity score matching.
(not needed for this PR, it;s more for extensions and reusing code for other matching tasks)